### PR TITLE
Adding logic to create forwarding rules

### PR DESCRIPTION
**Pull request by nikhiljindal**
Uses target proxies created in https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/pull/9 to create forwarding rules.

Limitations:
* Requires the user to have reserved a static IP and specified it using the static ip annotation.
* Can only create http forwarding rules. Need to add logic for https.

Will fix the limitations in future PRs.

cc @bowei @nicksardo @csbell @G-Harmon @madhusudancs

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/10)

### CODE REVIEW

**Comment by nikhiljindal** [comment]
The first commit is from https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/pull/9.
Will rebase to remove that commit once that PR merges so please ignore it while reviewing this PR.

**Comment by G-Harmon** [comment]
i can finish reviewing after #9 is submitted and it's easier to see the reset of the changes.

---

Reviewed 5 of 15 files at r1.
Review status: 5 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[app/mci/pkg/gcp/forwardingrule/forwardingrulesyncer.go, line 82 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-KxOq6EyB5LwwC1kHMkx:-KxOq6EyB5LwwC1kHMky:bqm6p47) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/d8ade4fd8eb575852068aea40a81ee4b6959433b/app/mci/pkg/gcp/forwardingrule/forwardingrulesyncer.go#L82)):*
> ```Go
> 	name := desiredFR.Name
> 	fmt.Println("Updating existing forwarding rule", name, "to match the desired state")
> 	if existingFR.IPAddress != desiredFR.IPAddress || existingFR.PortRange != desiredFR.PortRange {
> ```

Do you need to check protocol as well?

---

*[app/mci/pkg/gcp/forwardingrule/forwardingrulesyncer.go, line 90 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-KxOplzKQsj0Rr0aB5Co:-KxOplzKQsj0Rr0aB5Cp:bu6eud6) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/d8ade4fd8eb575852068aea40a81ee4b6959433b/app/mci/pkg/gcp/forwardingrule/forwardingrulesyncer.go#L90)):*
> if existingFR.IPAddress != desiredFR.IPAddress || existingFR.PortRange != desiredFR.PortRange {
> 		fmt.Println("Deleting the existing forwarding rule", name, "and will create a new one")
> 		if err := utils.IgnoreHTTPNotFound(s.frp.DeleteGlobalForwardingRule(name)); err != nil {
> 			return fmt.Errorf("error in deleting exising forwarding rule %s: %s", name, err)
> 		}
> 		return s.createForwardingRule(desiredFR)
> 	}
> 	// Update target proxy link in forwarding rule.
> 	err := s.frp.SetProxyForGlobalForwardingRule(name, desiredFR.Target)
Why do you have 2 branches, one for 'if forwardingRuleMatches' and another for checking IP+Range? Seems redundant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-:-KxOt2exCA9qD1VH3432:b-l4amrm)*

**Comment by nikhiljindal** [comment]
Rebased and pushed the latest code to remove commit from #9

**Comment by nikhiljindal** [comment]
Review status: 3 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[app/mci/pkg/gcp/forwardingrule/forwardingrulesyncer.go, line 82 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-KxOq6EyB5LwwC1kHMkx:-KxOvn_RpO3csFWXROwA:boj1qgt) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/d8ade4fd8eb575852068aea40a81ee4b6959433b/app/mci/pkg/gcp/forwardingrule/forwardingrulesyncer.go#L82)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

Do you need to check protocol as well?
</blockquote></details>

Good catch. Done.

---

*[app/mci/pkg/gcp/forwardingrule/forwardingrulesyncer.go, line 90 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-KxOplzKQsj0Rr0aB5Co:-KxOvs2B38ppiA4XKRrS:b-f34cyg) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/d8ade4fd8eb575852068aea40a81ee4b6959433b/app/mci/pkg/gcp/forwardingrule/forwardingrulesyncer.go#L90)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> if existingFR.IPAddress != desiredFR.IPAddress || existingFR.PortRange != desiredFR.PortRange {
> 		fmt.Println("Deleting the existing forwarding rule", name, "and will create a new one")
> 		if err := utils.IgnoreHTTPNotFound(s.frp.DeleteGlobalForwardingRule(name)); err != nil {
> 			return fmt.Errorf("error in deleting exising forwarding rule %s: %s", name, err)
> 		}
> 		return s.createForwardingRule(desiredFR)
> 	}
> 	// Update target proxy link in forwarding rule.
> 	err := s.frp.SetProxyForGlobalForwardingRule(name, desiredFR.Target)
Why do you have 2 branches, one for 'if forwardingRuleMatches' and another for checking IP+Range? Seems redundant.
</blockquote></details>

We do not have an UpdateForwardingRule method.
If Target proxy link is the only thing that is different, then we can call SetProxyForGlobalForwardingRule. Else, will need to delete the existing forwarding rule and create a new one.
Added a comment to this method to explain this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10)*

**Comment by G-Harmon** [comment]
:lgtm:

---

Reviewed 6 of 15 files at r1, 3 of 4 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 68 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-KxOvPxA2C0W-eZ8lHrK:-KxOvPxA2C0W-eZ8lHrL:brpp2dp) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/dab344c43b1e1d6a6eca4293b9441a18f40dffc9/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L68)):*
> ingress.global-static-ip-name
Can't you refer to kubernetes / ingress-gce / pkg / annotations / annotations.go now?

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 89 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-KxOva-qWIGCo_t0lqf-:-KxOva-qWIGCo_t0lqf0:b-svajq3) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/dab344c43b1e1d6a6eca4293b9441a18f40dffc9/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L89)):*
> ```Go
> 	igp ingressig.InstanceGroups
> 	// LoadBalancers interface to manage IP addresses.
> 	ipp ingresslb.LoadBalancers
> ```

what's ipp here? maybe reword the comment to make it clearer?

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go, line 65 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-KxOvx96-xCeOr7wQ9pA:-KxOvx96-xCeOr7wQ9pB:b-acx89b) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/dab344c43b1e1d6a6eca4293b9441a18f40dffc9/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go#L65)):*
> ```Go
> 	ipAddress := &compute.Address{
> 		Name:    "ipAddressName",
> 		Address: "192.168.0.0",
> ```

Use a valid address? (e.g. 192.168.0.1)

---

*[app/mci/pkg/gcp/targetproxy/targetproxysyncer.go, line 62 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-KxOwMlfPd4HyhPvxo1M:-KxOwMlfPd4HyhPvxo1N:bkq9uiz) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/dab344c43b1e1d6a6eca4293b9441a18f40dffc9/app/mci/pkg/gcp/targetproxy/targetproxysyncer.go#L62)):*
> ```Go
> // ensureHttpProxy ensures that the required target proxy exists for the given port.
> // Does nothing if it exists already, else creates a new one.
> func (s *TargetProxySyncer) ensureHttpProxy(lbName, umLink string) (string, error) {
> ```

mention the new return value

---

*[app/mci/pkg/gcp/targetproxy/targetproxysyncer_test.go, line 37 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-KxOwgwzoUfn2MUKz4vY:-KxOwgwzoUfn2MUKz4vZ:bt6x7j) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/dab344c43b1e1d6a6eca4293b9441a18f40dffc9/app/mci/pkg/gcp/targetproxy/targetproxysyncer_test.go#L37)):*
> ```Go
> 		t.Fatalf("expected NotFound error, got nil")
> 	}
> 	_, err := tps.EnsureTargetProxy(lbName, umLink)
> ```

check the target proxy link returned?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-:-KxOx5i17gMzTK7r4EZa:bnfp4nl)*

**Comment by nikhiljindal** [comment]
Review status: 11 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 68 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-KxOvPxA2C0W-eZ8lHrK:-KxOxe8K2rO0St_n-RZw:b4s83gl) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/dab344c43b1e1d6a6eca4293b9441a18f40dffc9/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L68)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> ingress.global-static-ip-name
Can't you refer to kubernetes / ingress-gce / pkg / annotations / annotations.go now?
</blockquote></details>

Doing that in https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/pull/11

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 89 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-KxOva-qWIGCo_t0lqf-:-KxOyEnyJxjvKny-H5hZ:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/dab344c43b1e1d6a6eca4293b9441a18f40dffc9/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L89)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

what's ipp here? maybe reword the comment to make it clearer?
</blockquote></details>

Done.

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go, line 65 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-KxOvx96-xCeOr7wQ9pA:-KxOz4dXG3_JRn1b7k2n:bs7bsk3) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/dab344c43b1e1d6a6eca4293b9441a18f40dffc9/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go#L65)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

Use a valid address? (e.g. 192.168.0.1)
</blockquote></details>

Why? 192.168.0.0 is a valid IP address? :)

---

*[app/mci/pkg/gcp/targetproxy/targetproxysyncer.go, line 62 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-KxOwMlfPd4HyhPvxo1M:-KxOyZAD84PXArg6xy7n:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/dab344c43b1e1d6a6eca4293b9441a18f40dffc9/app/mci/pkg/gcp/targetproxy/targetproxysyncer.go#L62)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

mention the new return value
</blockquote></details>

Done.

---

*[app/mci/pkg/gcp/targetproxy/targetproxysyncer_test.go, line 37 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10#-KxOwgwzoUfn2MUKz4vY:-KxOyy0x659fnm8rxxj6:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/dab344c43b1e1d6a6eca4293b9441a18f40dffc9/app/mci/pkg/gcp/targetproxy/targetproxysyncer_test.go#L37)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

check the target proxy link returned?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/10)*

**Comment by nikhiljindal** [comment]
Thanks @G-Harmon Updated code as per comments
